### PR TITLE
feat(cli): refactored setRole and unsetRole scripts to work with any …

### PR DESCRIPTION
**Pull request by xflord**
…role from perun-roles.yml

* the scripts now call getAllRolesManagementRules and hence are more flexible, since roles are no longer hard-coded.

### CODE REVIEW

File: perun-cli/setRole [hunk 1]
```diff
@@ -2,170 +2,125 @@
...
+		}
+		case "Resource" {
+			my $resourcesAgent = $agent->getResourcesAgent;
+			return $resourcesAgent->getResourceById( id => $objectId );
+		}
+		case "Service" {
+			my $servicesAgent = $agent->getServicesAgent;
+			return $servicesAgent->getServiceById( id => $objectId );
+		}
+		else {
```
Comment (Johaney-s) [inline perun-cli/setRole:122]: i think we will soon be working more also with securityAdmin role - worth adding it now to both scripts

File: perun-cli/Perun/beans/RoleManagementRules.pm [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+	if (defined($self->{_displayName})) {
+		$displayName = "$self->{_displayName}";
+	} else {
+		$displayName = undef;
+	}
+
+	return { roleName => $roleName, primaryObject => $primaryObject, privilegedRolesToManage => \@privilegedRolesToManage,
+		privilegedRolesToRead => \@privilegedRolesToRead, entitiesToManage => \%entitiesToManage, assignedObjects => \%assignedObjects,
+		assignmentCheck => \@assignmentCheck, associatedReadRoles => \@associatedReadRoles, assignableToAttributes => $assignableToAttributes,
+		skipMFA => $skipMFA, mfaCriticalRole => $mfaCriticalRole };
```
Comment (Johaney-s) [inline perun-cli/Perun/beans/RoleManagementRules.pm:109]:
```suggestion
		skipMFA => $skipMFA, mfaCriticalRole => $mfaCriticalRole, displayName => $displayName };
```

File: perun-cli/unsetRole [hunk 3]
```diff
@@ -2,173 +2,130 @@
...
 
-my $agent = Perun::Agent->new();
-my $authzResolverAgent = $agent->getAuthzResolverAgent;
-#Check options
 
-#one of userId or authorizedGroupId must be set
-if (!defined($userId) && !defined($authorizedGroupId)) { die "ERROR: one of userId and authorizedGroupId is requried\n";}
-if (defined($userId) && defined($authorizedGroupId)) { die "ERROR: only one of userId and authorizedGroupId must be set at one moment\n";}
+# One of userId or authorizedGroupId must be set
+if (!defined($userId) && !defined($authorizedGroupId)) { die "ERROR: one of userId and authorizedGroupId is requried\n"; }
```
Comment (Johaney-s) [inline perun-cli/unsetRole:45]:
```suggestion
if (!defined($userId) && !defined($authorizedGroupId)) { die "ERROR: one of userId and authorizedGroupId is required\n"; }
```

File: perun-cli/setRole [hunk 4]
```diff
@@ -2,170 +2,125 @@
...
 
-my $agent = Perun::Agent->new();
-my $authzResolverAgent = $agent->getAuthzResolverAgent;
-#Check options
 
-#one of userId or authorizedGroupId must be set
-if (!defined($userId) && !defined($authorizedGroupId)) { die "ERROR: one of userId and authorizedGroupId is requried\n";}
-if (defined($userId) && defined($authorizedGroupId)) { die "ERROR: only one of userId and authorizedGroupId must be set at one moment\n";}
+# One of userId or authorizedGroupId must be set
+if (!defined($userId) && !defined($authorizedGroupId)) { die "ERROR: one of userId and authorizedGroupId is requried\n"; }
```
Comment (Johaney-s) [inline perun-cli/setRole:45]:
```suggestion
if (!defined($userId) && !defined($authorizedGroupId)) { die "ERROR: one of userId and authorizedGroupId is required\n"; }
```

File: perun-cli/unsetRole [hunk 5]
```diff
@@ -2,173 +2,130 @@
...
-	} elsif (!defined($facilityId) && defined($groupId) && !defined($voId) && !defined($resourceId) && !defined($serviceId)) {
-		my $groupsAgent = $agent->getGroupsAgent;
-		my $group = $groupsAgent->getGroupById( id => $groupId );
-		if (defined($userId)) {
-			$authzResolverAgent->unsetRole( user => $userId, complementaryObject => $group, role => $role );
-			printMessage("$role role was removed from user $userId and group $groupId", $batch);
-		} else {
-			$authzResolverAgent->unsetRole( authorizedGroup => $authorizedGroupId, complementaryObject => $group, role => $role );
-			printMessage("$role role was removed from authorizedGroup $authorizedGroupId and group $groupId", $batch);
+# # Complementary object must be set
```
Comment (Johaney-s) [inline perun-cli/unsetRole:52]: this block of comments can be removed, right?

File: perun-cli/unsetRole [hunk 6]
```diff
@@ -2,173 +2,130 @@
...
+
+foreach (@rules) {
+	my $rule = $_;
+	if ($rule->getRoleName() eq $role) {
+		my $primObject = $rule->getPrimaryObject();
+		if (!defined($primObject)) {
+			if (defined($compObjectName)) {
+				die "ERROR: role $role doesn't have a complementary object\n";
+			}
+		} elsif (!defined($compObjectName) || !defined($objectId) || $primObject ne $compObjectName) {
```
Comment (Johaney-s) [inline perun-cli/unsetRole:70]: a nitpick - would be nice to have the complementary object check case insensitive

**Review by Johaney-s — COMMENTED** [review]
superb, just a few suggestions

File: perun-cli/setRole [hunk 7]
```diff
@@ -2,170 +2,125 @@
...
 use Perun::Common qw(printMessage);
 
+
 sub help {
 	return qq{
-	Adds Role to user or authorizedGroup. User or group is required.
-	Role is required. Some complementaryObject is required if not PERUNADMIN.
-	For now only 1 object is supported to be set at once.
+	Adds role to user or authorizedGroup. User or group is required.
+	Name of the role to set in all caps with no spaces is required (e.g. GROUPADMIN).
```
Comment (mattjoke) [inline perun-cli/setRole:14]: A little nit-pick.
```suggestion
Name of the role set in all caps with no spaces is required
```

**Review by mattjoke — APPROVED** [review]
Apart for the little things mentioned above I think it looks good :+1:

**Review by Johaney-s — APPROVED** [review]

**Pull request merged**